app/models/models.py

```diff
@@ -0,0 +1,151 @@
+from datetime import datetime, timedelta
+from decimal import Decimal
+from math import floor
+from typing import Any, Dict, List, Set
+
+from config import (
+	DATE_TIME_FORMAT,
+	FREE_MINUTES,
+	TIME_FRAME_HOURS,
+	TIME_FRAME_TARIFFS
+)
+
+class ParkingLotFullError(Exception):
+	"""Custom error that is raised when the Parking Lot is full"""
+	pass
+
+class ParkingLotLocationEmptyError(Exception):
+	"""Custom error that is raised when the Parking Lot location is empty"""
+	pass
+
+class Ticket:
+	def __init__(self, license_plate: str, tariff: str, location: int) -> None:
+		"""Ticket constructor
+
+		Args:
+			license_plate (str): Car license plate.
+			tariff (str): Tariff for the ticket.
+			location (int): Parking Lot location.
+
+		Returns:
+			None
+		"""
+		self.car: str = license_plate
+		self.tariff: str = tariff
+		self.location: int = location
+		self.start: str = datetime.now().strftime(DATE_TIME_FORMAT)
+		self.finish: str = ""
+		self.fee: str = ""
+
+	def _get_total_time(self) -> timedelta:
+		"""Get the total time for the ticket
+
+		Returns:
+			timedelta: Total time for the ticket.
+		"""
+		try:
+			start = datetime.strptime(self.start, DATE_TIME_FORMAT)
+			finish = datetime.strptime(self.finish, DATE_TIME_FORMAT)
+		except ValueError as e:
+			raise ValueError("Invalid date format") from e
+		return finish - start
+
+	def calculate_fee(self, end_datetime: datetime = datetime.now()) -> None:
+		"""Calculate the fee for the ticket
+
+		Returns:
+			None
+		"""
+		self.finish = end_datetime.strftime(DATE_TIME_FORMAT)
+		delta = self._get_total_time()
+		total_minutes = delta / timedelta(minutes=1)
+
+		if total_minutes <= FREE_MINUTES:
+			self.fee = "0.00"
+			return
+
+		time_frame = TIME_FRAME_HOURS.get(self.tariff)
+		tariff = TIME_FRAME_TARIFFS.get(self.tariff)
+
+		total_time = delta / timedelta(hours=time_frame)
+
+		if not total_time.is_integer():
+			total_time += 1
+
+		self.fee = str(Decimal(tariff) * Decimal(floor(total_time)))
```

Comment:
What changes are necessary to support new tariffs dynamically ?

Reply:
The Parking Lot owner would need an API endpoint to change the tariffs as wanted.

I implemented a /config endpoint to change the parking lot tariffs using a POST request. If the owner wants to know the current tariffs without adding a car, they can be retrieved with the same endpoint by using a GET request instead.